logs-to-tsv/query-logs.js

```diff
@@ -0,0 +1,83 @@
+const { CloudWatchLogsClient, StartQueryCommand, GetQueryResultsCommand } = require("@aws-sdk/client-cloudwatch-logs")
+const { writeFileSync } = require('fs')
+const aws = require('aws-sdk')
+
+const awsInit = (profile) => {
+  // Set aws creds:
+  aws.config.credentials = new aws.SharedIniFileCredentials({
+    profile: profile || 'nypl-digital-dev'
+  })
+
+  // Set aws region:
+  const awsSecurity = { region: 'us-east-1' }
+  aws.config.update(awsSecurity)
+}
+
+awsInit()
+
+const client = new CloudWatchLogsClient();
+
+const start = new Date();
+start.setMonth(start.getMonth() - 3)
+
+const inputs = []
+for (let i = 0; i < 20; i++) {
+  inputs.push({
+    "startTime": Date.parse(start),
+    "endTime": start.setDate(start.getDate() + 2),
+    "logGroupName": "/aws/elasticbeanstalk/discovery-api-production/var/log/nginx/access.log",
+    "queryString": "fields @message | sort @timestamp desc | filter @message like \"/resources?\" | filter @message not like \"filters\" | limit 10000 | filter @message not like \"oclc\"",
+  })
+}
+
+const delay = async (time) => new Promise((resolve) => setTimeout(resolve, time))
+
+const checkResultStatusAndDownload = async (queryIds, completedQueries = {}, numberOfChecks = 1) => {
+  while (queryIds.length !== Object.keys(completedQueries).length) {
```

Comment:
I don't think a `while` is necessary here. It implies the script is continually checking the condition, but in fact you `return` on the first pass, so it functions as an `if`. Also, minor thing, but the placement of the `delay` means the script is going to wait an extra 10s even if everything is complete. I think you can resolve both issues if the structure was:
```
checkResultStatusAndDownload = async (queryIds, completedQueries = {}, numberOfChecks = 1) => {
  await Promise.all(queryIds.map(async (queryId, i) => {
    // Have we already downloaded this report? Skip:
    if (completedQueries[queryId]) return Promise.resolve()
    
    // Do check for completed query here...
  }))
  
  // Are there more to download? Wait 10s and recurse:  
  if (queryIds.length !== Object.keys(completedQueries).length) {
    await delay(10000)
    return await checkResultStatusAndDownload(queryIds, completedQueries, ++numberOfChecks)
  }
}
```